File: src/main/java/no/nav/fo/veilarbregistrering/registrering/bruker/db/BrukerRegistreringRepositoryImpl.java
package no.nav.fo.veilarbregistrering.registrering.bruker.db;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import no.nav.fo.veilarbregistrering.bruker.AktorId;
import no.nav.fo.veilarbregistrering.registrering.bruker.*;
import no.nav.fo.veilarbregistrering.besvarelse.*;
import no.nav.sbl.sql.DbConstants;
import no.nav.sbl.sql.SqlUtils;
import no.nav.sbl.sql.order.OrderClause;
import no.nav.sbl.sql.where.WhereClause;
import org.springframework.jdbc.core.JdbcTemplate;

import java.io.IOException;
import java.sql.Timestamp;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;

import static java.util.Optional.ofNullable;

public class BrukerRegistreringRepositoryImpl implements BrukerRegistreringRepository {

    private JdbcTemplate db;

    private final static String SYKMELDT_REGISTRERING_SEQ = "SYKMELDT_REGISTRERING_SEQ";
    final static String SYKMELDT_REGISTRERING_ID = "SYKMELDT_REGISTRERING_ID";
    private final static String SYKMELDT_REGISTRERING = "SYKMELDT_REGISTRERING";
    final static String FREMTIDIG_SITUASJON = "FREMTIDIG_SITUASJON";
    final static String TILBAKE_ETTER_52_UKER = "TILBAKE_ETTER_52_UKER";

    private final static String REGISTRERING_TILSTAND_SEQ = "REGISTRERING_TILSTAND_SEQ";

    private final static String BRUKER_REGISTRERING_SEQ = "BRUKER_REGISTRERING_SEQ";
    private final static String BRUKER_REAKTIVERING_SEQ = "BRUKER_REAKTIVERING_SEQ";
    private final static String BRUKER_REGISTRERING = "BRUKER_REGISTRERING";
    private final static String BRUKER_REAKTIVERING = "BRUKER_REAKTIVERING";
    final static String BRUKER_REGISTRERING_ID = "BRUKER_REGISTRERING_ID";
    private final static String BRUKER_REAKTIVERING_ID = "BRUKER_REAKTIVERING_ID";
    final static String OPPRETTET_DATO = "OPPRETTET_DATO";
    private final static String REAKTIVERING_DATO = "REAKTIVERING_DATO";

    final static String NUS_KODE = "NUS_KODE";
    final static String YRKESPRAKSIS = "YRKESPRAKSIS";
    final static String HAR_HELSEUTFORDRINGER = "HAR_HELSEUTFORDRINGER";
    final static String YRKESBESKRIVELSE = "YRKESBESKRIVELSE";
    final static String KONSEPT_ID = "KONSEPT_ID";
    final static String TEKSTER_FOR_BESVARELSE = "TEKSTER_FOR_BESVARELSE";

    final static String ANDRE_UTFORDRINGER = "ANDRE_UTFORDRINGER";
    final static String BEGRUNNELSE_FOR_REGISTRERING = "BEGRUNNELSE_FOR_REGISTRERING";
    final static String UTDANNING_BESTATT = "UTDANNING_BESTATT";
    final static String UTDANNING_GODKJENT_NORGE = "UTDANNING_GODKJENT_NORGE";
    final static String JOBBHISTORIKK = "JOBBHISTORIKK";

    private final static String AKTOR_ID = "AKTOR_ID";

    public BrukerRegistreringRepositoryImpl(JdbcTemplate db) {
        this.db = db;
    }

    @Override
    public OrdinaerBrukerRegistrering lagreOrdinaerBruker(OrdinaerBrukerRegistrering bruker, AktorId aktorId) {
        long id = nesteFraSekvens(BRUKER_REGISTRERING_SEQ);
        Besvarelse besvarelse = bruker.getBesvarelse();
        Stilling stilling = bruker.getSisteStilling();
        String teksterForBesvarelse = tilJson(bruker.getTeksterForBesvarelse());

        SqlUtils.insert(db, BRUKER_REGISTRERING)
                .value(BRUKER_REGISTRERING_ID, id)
                .value(AKTOR_ID, aktorId.asString())
                .value(OPPRETTET_DATO, DbConstants.CURRENT_TIMESTAMP)
                .value(TEKSTER_FOR_BESVARELSE, teksterForBesvarelse)
                // Siste stilling
                .value(YRKESPRAKSIS, stilling.getStyrk08())
                .value(YRKESBESKRIVELSE, stilling.getLabel())
                .value(KONSEPT_ID, stilling.getKonseptId())
                // Besvarelse
                .value(BEGRUNNELSE_FOR_REGISTRERING, besvarelse.getDinSituasjon().toString())
                .value(NUS_KODE, UtdanningUtils.mapTilNuskode(besvarelse.getUtdanning()))
                .value(UTDANNING_GODKJENT_NORGE, besvarelse.getUtdanningGodkjent().toString())
                .value(UTDANNING_BESTATT, besvarelse.getUtdanningBestatt().toString())
                .value(HAR_HELSEUTFORDRINGER, besvarelse.getHelseHinder().toString())
                .value(ANDRE_UTFORDRINGER, besvarelse.getAndreForhold().toString())
                .value(JOBBHISTORIKK, besvarelse.getSisteStilling().toString())
                .execute();

        return hentBrukerregistreringForId(id);
    }

    @Override
    public long lagreSykmeldtBruker(SykmeldtRegistrering bruker, AktorId aktorId) {
        long id = nesteFraSekvens(SYKMELDT_REGISTRERING_SEQ);
        Besvarelse besvarelse = bruker.getBesvarelse();
        String teksterForBesvarelse = tilJson(bruker.getTeksterForBesvarelse());

        SqlUtils.insert(db, SYKMELDT_REGISTRERING)
                .value(SYKMELDT_REGISTRERING_ID, id)
                .value(AKTOR_ID, aktorId.asString())
                .value(OPPRETTET_DATO, DbConstants.CURRENT_TIMESTAMP)
                .value(TEKSTER_FOR_BESVARELSE, teksterForBesvarelse)
                // Besvarelse
                .value(FREMTIDIG_SITUASJON, ofNullable(besvarelse.getFremtidigSituasjon()).map(Enum::toString).orElse(null))
                .value(TILBAKE_ETTER_52_UKER, ofNullable(besvarelse.getTilbakeIArbeid()).map(Enum::toString).orElse(null))
                .value(NUS_KODE, ofNullable(UtdanningUtils.mapTilNuskode(besvarelse.getUtdanning())).orElse(null))
                .value(UTDANNING_BESTATT, ofNullable(besvarelse.getUtdanningBestatt()).map(Enum::toString).orElse(null))
                .value(UTDANNING_GODKJENT_NORGE, ofNullable(besvarelse.getUtdanningGodkjent()).map(Enum::toString).orElse(null))
                .value(ANDRE_UTFORDRINGER, ofNullable(besvarelse.getAndreForhold()).map(Enum::toString).orElse(null))
                .execute();

        return id;
    }

    private static String tilJson(List<TekstForSporsmal> obj) {
        if (obj == null) {
            return "[]";
        }
        try {
            return (new ObjectMapper()).writeValueAsString(obj);
        } catch (JsonProcessingException e) {
            return "[]";
        }
    }

    static List<TekstForSporsmal> tilTeksterForBesvarelse(String json) {
        try {
            TekstForSporsmal[] teksterForBesvarelse = (new ObjectMapper()).readValue(json, TekstForSporsmal[].class);
            return teksterForBesvarelse != null ? Arrays.asList(teksterForBesvarelse) : new ArrayList<>();
        } catch (IOException e) {
            return new ArrayList<>();
        }
    }

    @Override
    public OrdinaerBrukerRegistrering hentBrukerregistreringForId(long brukerregistreringId) {
        return SqlUtils.select(db, BRUKER_REGISTRERING, OrdinaerBrukerRegistreringMapper::map)
                .where(WhereClause.equals(BRUKER_REGISTRERING_ID, brukerregistreringId))
                .column("*")
                .execute();
    }

    @Override
    public OrdinaerBrukerRegistrering hentOrdinaerBrukerregistreringForAktorId(AktorId aktorId) {
        return SqlUtils.select(db, BRUKER_REGISTRERING, OrdinaerBrukerRegistreringMapper::map)
                .where(WhereClause.equals(AKTOR_ID, aktorId.asString()))
                .orderBy(OrderClause.desc(BRUKER_REGISTRERING_ID))
                .limit(1)
                .column("*")
                .execute();
    }

    @Override
    public SykmeldtRegistrering hentSykmeldtregistreringForAktorId(AktorId aktorId) {
        return SqlUtils.select(db, SYKMELDT_REGISTRERING, SykmeldtRegistreringMapper::map)
                .where(WhereClause.equals(AKTOR_ID, aktorId.asString()))
                .orderBy(OrderClause.desc(SYKMELDT_REGISTRERING_ID))
                .limit(1)
                .column("*")
                .execute();
    }

    @Override
    public void lagreReaktiveringForBruker(AktorId aktorId) {
        long id = nesteFraSekvens(BRUKER_REAKTIVERING_SEQ);
        SqlUtils.insert(db, BRUKER_REAKTIVERING)
                .value(BRUKER_REAKTIVERING_ID, id)
                .value(AKTOR_ID, aktorId.asString())
                .value(REAKTIVERING_DATO, DbConstants.CURRENT_TIMESTAMP)
                .execute();
    }

    @Override
    public long lagre(RegistreringTilstand registreringTilstand) {
        long id = nesteFraSekvens(REGISTRERING_TILSTAND_SEQ);
        SqlUtils.insert(db, "REGISTRERING_TILSTAND")
                .value("ID", id)
                .value("UUID", registreringTilstand.getUuid().toString())
                .value("BRUKER_REGISTRERING_ID", registreringTilstand.getBrukerRegistreringId())
                .value("OPPRETTET", Timestamp.valueOf(registreringTilstand.getOpprettet()))
                .value("SIST_ENDRET", ofNullable(registreringTilstand.getSistEndret())
                        .map(Timestamp::valueOf).orElse(null))
                .value("STATUS", registreringTilstand.getStatus().toString())
                .execute();

        return id;
    }

    @Override
    public void oppdater(RegistreringTilstand registreringTilstand) {
        SqlUtils.update(db, "REGISTRERING_TILSTAND")
                .set("STATUS", registreringTilstand.getStatus())
                .set("SIST_ENDRET", Timestamp.valueOf(registreringTilstand.getSistEndret()))
                .whereEquals("ID", registreringTilstand.getId())
                .execute();
    }

    @Override
    public RegistreringTilstand hentRegistreringTilstand(long id) {
        return SqlUtils.select(db, "REGISTRERING_TILSTAND", RegistreringTilstandMapper::map)
                .where(WhereClause.equals("ID", id))
                .column("*")
                .execute();
    }

    @Override
    public Optional<RegistreringTilstand> finnNesteRegistreringForOverforing() {
        RegistreringTilstand registreringTilstand = SqlUtils.select(db, "REGISTRERING_TILSTAND", RegistreringTilstandMapper::map)
                .where(WhereClause.equals("STATUS", "MOTTATT"))
                .orderBy(OrderClause.asc("OPPRETTET"))
                .limit(1)
                .column("*")
                .execute();

        return ofNullable(registreringTilstand);
    }

    private long nesteFraSekvens(String sekvensNavn) {
        return ((Long)this.db.queryForObject("select " + sekvensNavn + ".nextval from dual", Long.class)).longValue();
    }

}

File: src/main/kotlin/no/nav/fo/veilarbregistrering/registrering/ordinaer/db/BrukerRegistreringRepositoryImpl.kt
package no.nav.fo.veilarbregistrering.registrering.ordinaer.db

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import no.nav.fo.veilarbregistrering.besvarelse.AndreForholdSvar
import no.nav.fo.veilarbregistrering.besvarelse.Besvarelse
import no.nav.fo.veilarbregistrering.besvarelse.DinSituasjonSvar
import no.nav.fo.veilarbregistrering.besvarelse.HelseHinderSvar
import no.nav.fo.veilarbregistrering.besvarelse.SisteStillingSvar
import no.nav.fo.veilarbregistrering.besvarelse.Stilling
import no.nav.fo.veilarbregistrering.besvarelse.UtdanningBestattSvar
import no.nav.fo.veilarbregistrering.besvarelse.UtdanningGodkjentSvar
import no.nav.fo.veilarbregistrering.besvarelse.UtdanningSvar
import no.nav.fo.veilarbregistrering.besvarelse.UtdanningUtils
import no.nav.fo.veilarbregistrering.bruker.AktorId
import no.nav.fo.veilarbregistrering.bruker.Bruker
import no.nav.fo.veilarbregistrering.bruker.Foedselsnummer
import no.nav.fo.veilarbregistrering.registrering.bruker.TekstForSporsmal
import no.nav.fo.veilarbregistrering.registrering.formidling.Status
import no.nav.fo.veilarbregistrering.registrering.formidling.db.RegistreringTilstandRepositoryImpl
import no.nav.fo.veilarbregistrering.registrering.formidling.db.RegistreringTilstandRepositoryImpl.Companion.REGISTRERING_TILSTAND
import no.nav.fo.veilarbregistrering.registrering.ordinaer.BrukerRegistreringRepository
import no.nav.fo.veilarbregistrering.registrering.ordinaer.OrdinaerBrukerRegistrering
import no.nav.paw.arbeidssokerregisteret.intern.v1.OpplysningerOmArbeidssoekerMottatt
import no.nav.paw.arbeidssokerregisteret.intern.v1.vo.Annet
import no.nav.paw.arbeidssokerregisteret.intern.v1.vo.Arbeidserfaring
import no.nav.paw.arbeidssokerregisteret.intern.v1.vo.BrukerType
import no.nav.paw.arbeidssokerregisteret.intern.v1.vo.Helse
import no.nav.paw.arbeidssokerregisteret.intern.v1.vo.JaNeiVetIkke
import no.nav.paw.arbeidssokerregisteret.intern.v1.vo.Jobbsituasjon
import no.nav.paw.arbeidssokerregisteret.intern.v1.vo.JobbsituasjonBeskrivelse
import no.nav.paw.arbeidssokerregisteret.intern.v1.vo.JobbsituasjonMedDetaljer
import no.nav.paw.arbeidssokerregisteret.intern.v1.vo.Metadata
import no.nav.paw.arbeidssokerregisteret.intern.v1.vo.OpplysningerOmArbeidssoeker
import no.nav.paw.arbeidssokerregisteret.intern.v1.vo.Utdanning
import no.nav.paw.arbeidssokerregisteret.intern.v1.vo.Utdanningsnivaa
import org.springframework.jdbc.core.RowMapper
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate
import java.io.IOException
import java.sql.SQLException
import java.sql.Timestamp
import java.util.*

class BrukerRegistreringRepositoryImpl(private val db: NamedParameterJdbcTemplate) : BrukerRegistreringRepository {

    override fun lagre(registrering: OrdinaerBrukerRegistrering, bruker: Bruker): OrdinaerBrukerRegistrering {
        val id = nesteFraSekvens(BRUKER_REGISTRERING_SEQ)
        val besvarelse = registrering.besvarelse
        val stilling = registrering.sisteStilling
        val teksterForBesvarelse = tilJson(registrering.teksterForBesvarelse)

        val params = mapOf(
            "id" to id,
            "aktor_id" to bruker.aktorId.aktorId,
            "fnr" to bruker.gjeldendeFoedselsnummer.stringValue(),
            "opprettet" to Timestamp.valueOf(registrering.opprettetDato),
            "tekster" to teksterForBesvarelse,
            "yrkespraksis" to stilling.styrk08,
            "yrkesbeskrivelse" to stilling.label,
            "konsept_id" to stilling.konseptId,
            "begrunnelse" to besvarelse.dinSituasjon.toString(),
            "nus_kode" to UtdanningUtils.mapTilNuskode(besvarelse.utdanning),
            "utdanning_godkjent" to besvarelse.utdanningGodkjent.toString(),
            "utdanning_bestatt" to besvarelse.utdanningBestatt.toString(),
            "har_helseutfordringer" to besvarelse.helseHinder.toString(),
            "andre_utfordringer" to besvarelse.andreForhold.toString(),
            "jobbhistorikk" to besvarelse.sisteStilling.toString()
        )
        val sql = "INSERT INTO $BRUKER_REGISTRERING " +
                " (${allColumns.joinToString(", ")})" +
                " VALUES (:id, :aktor_id, :fnr, :opprettet, :tekster, :yrkespraksis, :yrkesbeskrivelse," +
                " :konsept_id, :begrunnelse, :nus_kode, :utdanning_godkjent, :utdanning_bestatt," +
                " :har_helseutfordringer, :andre_utfordringer, :jobbhistorikk)"

        db.update(sql, params)
        return hentBrukerregistreringForId(id)
    }

    override fun hentBrukerregistreringForId(brukerregistreringId: Long): OrdinaerBrukerRegistrering {
        val sql = "SELECT * FROM $BRUKER_REGISTRERING WHERE $BRUKER_REGISTRERING_ID = :id"

        return db.queryForObject(sql, mapOf("id" to brukerregistreringId), registreringMapper)!!
    }

    override fun hentBrukerregistreringForFoedselsnummer(foedselsnummerList: List<Foedselsnummer>): List<OrdinaerBrukerRegistrering> {
        val sql = "SELECT * FROM $BRUKER_REGISTRERING" +
                " LEFT JOIN $REGISTRERING_TILSTAND ON" +
                " $REGISTRERING_TILSTAND.${RegistreringTilstandRepositoryImpl.BRUKER_REGISTRERING_ID} = $BRUKER_REGISTRERING.$BRUKER_REGISTRERING_ID" +
                " WHERE $FOEDSELSNUMMER IN (:foedselsnummer)" +
                " AND $REGISTRERING_TILSTAND.${RegistreringTilstandRepositoryImpl.STATUS} in (:tilstander)"

        val parameters = mapOf(
            "foedselsnummer" to foedselsnummerList.map(Foedselsnummer::stringValue),
            "tilstander" to Status.gyldigTilstand().map { it.name }
        )

        return db.query(sql, parameters, registreringMapper)
    }

    override fun finnOrdinaerBrukerregistreringForAktorIdOgTilstand(
        aktorId: AktorId,
        tilstander: List<Status>
    ): List<OrdinaerBrukerRegistrering> {
        val params = mapOf(
            "aktor_id" to aktorId.aktorId,
            "tilstander" to tilstander.map { it.name }
        )

        val sql = "SELECT * FROM $BRUKER_REGISTRERING" +
                " LEFT JOIN $REGISTRERING_TILSTAND ON " +
                " $REGISTRERING_TILSTAND.${RegistreringTilstandRepositoryImpl.BRUKER_REGISTRERING_ID} = $BRUKER_REGISTRERING.$BRUKER_REGISTRERING_ID" +
                " WHERE $BRUKER_REGISTRERING.$AKTOR_ID = :aktor_id" +
                " AND $REGISTRERING_TILSTAND.${RegistreringTilstandRepositoryImpl.STATUS} in (:tilstander)" +
                " ORDER BY $BRUKER_REGISTRERING.$OPPRETTET_DATO DESC"

        return db.query(sql, params, registreringMapper)
    }

    override fun hentBrukerTilknyttet(brukerRegistreringId: Long): Bruker {
        val sql = "SELECT $FOEDSELSNUMMER, $AKTOR_ID FROM $BRUKER_REGISTRERING WHERE $BRUKER_REGISTRERING_ID = :id"

        return db.queryForObject(sql, mapOf("id" to brukerRegistreringId)) { rs, _ ->
            Bruker(
                Foedselsnummer(rs.getString("FOEDSELSNUMMER")),
                AktorId(rs.getString("AKTOR_ID"))
            )
        }!!
    }

    override fun hentNesteOpplysningerOmArbeidssoeker(antall: Int): List<Pair<Long, OpplysningerOmArbeidssoekerMottatt>> {
        val sql = """
            SELECT bruker_registrering.* FROM $BRUKER_REGISTRERING
            JOIN registrering_tilstand ON bruker_registrering.bruker_registrering_id = registrering_tilstand.bruker_registrering_id
            WHERE registrering_tilstand.status in ('PUBLISERT_KAFKA','OPPRINNELIG_OPPRETTET_UTEN_TILSTAND')
            AND overfort_kafka IS FALSE
            AND $FOEDSELSNUMMER IS NOT NULL
            ORDER BY $OPPRETTET_DATO LIMIT $antall
        """.trimIndent()

        return db.query(sql, noParams, opplysningerOmArbeidssoekerMapper)
    }

    override fun settOpplysningerOmArbeidssoekerSomOverfort(listeMedIder: List<Int>) {
        val sql =
            "UPDATE $BRUKER_REGISTRERING SET overfort_kafka = TRUE WHERE $BRUKER_REGISTRERING_ID IN (:listeMedIder)"
        val params = mapOf("listeMedIder" to listeMedIder)

        db.update(sql, params)
    }


    private fun nesteFraSekvens(sekvensNavn: String): Long {
        val sql = "SELECT nextVal('$sekvensNavn')"
        return db.queryForObject(sql, noParams, Long::class.java)!!
    }

    companion object {
        private val mapper = jacksonObjectMapper()

        private const val BRUKER_REGISTRERING = "BRUKER_REGISTRERING"
        private const val BRUKER_REGISTRERING_SEQ = "BRUKER_REGISTRERING_SEQ"
        const val FREMTIDIG_SITUASJON = "FREMTIDIG_SITUASJON"
        const val TILBAKE_ETTER_52_UKER = "TILBAKE_ETTER_52_UKER"
        const val BRUKER_REGISTRERING_ID = "BRUKER_REGISTRERING_ID"
        const val OPPRETTET_DATO = "OPPRETTET_DATO"
        const val NUS_KODE = "NUS_KODE"
        const val YRKESPRAKSIS = "YRKESPRAKSIS"
        private const val HAR_HELSEUTFORDRINGER = "HAR_HELSEUTFORDRINGER"
        const val YRKESBESKRIVELSE = "YRKESBESKRIVELSE"
        const val KONSEPT_ID = "KONSEPT_ID"
        const val TEKSTER_FOR_BESVARELSE = "TEKSTER_FOR_BESVARELSE"
        const val ANDRE_UTFORDRINGER = "ANDRE_UTFORDRINGER"
        const val BEGRUNNELSE_FOR_REGISTRERING = "BEGRUNNELSE_FOR_REGISTRERING"
        const val UTDANNING_BESTATT = "UTDANNING_BESTATT"
        const val UTDANNING_GODKJENT_NORGE = "UTDANNING_GODKJENT_NORGE"
        private const val JOBBHISTORIKK = "JOBBHISTORIKK"
        private const val FOEDSELSNUMMER = "FOEDSELSNUMMER"
        const val AKTOR_ID = "AKTOR_ID"

        private val allColumns = listOf(
            BRUKER_REGISTRERING_ID,
            AKTOR_ID,
            FOEDSELSNUMMER,
            OPPRETTET_DATO,
            TEKSTER_FOR_BESVARELSE,
            YRKESPRAKSIS,
            YRKESBESKRIVELSE,
            KONSEPT_ID,
            BEGRUNNELSE_FOR_REGISTRERING,
            NUS_KODE,
            UTDANNING_GODKJENT_NORGE,
            UTDANNING_BESTATT,
            HAR_HELSEUTFORDRINGER,
            ANDRE_UTFORDRINGER,
            JOBBHISTORIKK
        )

        private val registreringMapper = RowMapper<OrdinaerBrukerRegistrering> { rs, _ ->
            try {
                OrdinaerBrukerRegistrering(
                    id = rs.getLong(BRUKER_REGISTRERING_ID),
                    opprettetDato = rs.getTimestamp(OPPRETTET_DATO).toLocalDateTime(),
                    teksterForBesvarelse = readListOf(rs.getString(TEKSTER_FOR_BESVARELSE)),
                    sisteStilling =
                    Stilling(
                        styrk08 = rs.getString(YRKESPRAKSIS),
                        konseptId = rs.getLong(KONSEPT_ID),
                        label = rs.getString(YRKESBESKRIVELSE)
                    ),
                    besvarelse = (
                            Besvarelse(
                                dinSituasjon = DinSituasjonSvar.valueOf(rs.getString(BEGRUNNELSE_FOR_REGISTRERING)),
                                utdanning = UtdanningUtils.mapTilUtdanning(rs.getString(NUS_KODE)),
                                utdanningBestatt = UtdanningBestattSvar.valueOf(rs.getString(UTDANNING_BESTATT)),
                                utdanningGodkjent = UtdanningGodkjentSvar.valueOf(rs.getString(UTDANNING_GODKJENT_NORGE)),
                                helseHinder = HelseHinderSvar.valueOf(rs.getString(HAR_HELSEUTFORDRINGER)),
                                andreForhold = AndreForholdSvar.valueOf(rs.getString(ANDRE_UTFORDRINGER)),
                                sisteStilling = SisteStillingSvar.valueOf(rs.getString(JOBBHISTORIKK)),
                            )
                            ),
                )
            } catch (e: SQLException) {
                throw RuntimeException(e)
            }
        }

        val opplysningerOmArbeidssoekerMapper = RowMapper<Pair<Long, OpplysningerOmArbeidssoekerMottatt>> { rs, _ ->
            rs.getLong(BRUKER_REGISTRERING_ID) to OpplysningerOmArbeidssoekerMottatt(
                hendelseId = UUID.randomUUID(),
                identitetsnummer = rs.getString(FOEDSELSNUMMER),
                opplysningerOmArbeidssoeker = OpplysningerOmArbeidssoeker(
                    id = UUID.randomUUID(),
                    metadata = Metadata(
                        tidspunkt = rs.getTimestamp(OPPRETTET_DATO).toInstant(),
                        utfoertAv = no.nav.paw.arbeidssokerregisteret.intern.v1.vo.Bruker(
                            type = BrukerType.SLUTTBRUKER,
                            id = rs.getString(FOEDSELSNUMMER)
                        ),
                        kilde = "veilarbregistrering",
                        aarsak = "registrering"
                    ),
                    utdanning = Utdanning(
                        nus = rs.getString(NUS_KODE),
                        bestaatt = when (UtdanningBestattSvar.valueOf(rs.getString(UTDANNING_BESTATT))) {
                            UtdanningBestattSvar.JA -> JaNeiVetIkke.JA
                            UtdanningBestattSvar.NEI -> JaNeiVetIkke.NEI
                            else -> JaNeiVetIkke.VET_IKKE
                        },
                        godkjent = when (UtdanningGodkjentSvar.valueOf(rs.getString(UTDANNING_GODKJENT_NORGE))) {
                            UtdanningGodkjentSvar.JA -> JaNeiVetIkke.JA
                            UtdanningGodkjentSvar.NEI -> JaNeiVetIkke.NEI
                            UtdanningGodkjentSvar.VET_IKKE -> JaNeiVetIkke.VET_IKKE
                            UtdanningGodkjentSvar.INGEN_SVAR -> JaNeiVetIkke.VET_IKKE
                        }
                    ),
                    arbeidserfaring = Arbeidserfaring(
                        harHattArbeid = when (SisteStillingSvar.valueOf(rs.getString(JOBBHISTORIKK))) {
                            SisteStillingSvar.HAR_HATT_JOBB -> JaNeiVetIkke.JA
                            SisteStillingSvar.HAR_IKKE_HATT_JOBB -> JaNeiVetIkke.NEI
                            SisteStillingSvar.INGEN_SVAR -> JaNeiVetIkke.VET_IKKE
                        }
                    ),
                    jobbsituasjon = Jobbsituasjon(
                        beskrivelser = listOfNotNull(
                            when (DinSituasjonSvar.valueOf(rs.getString(BEGRUNNELSE_FOR_REGISTRERING))) {
                                DinSituasjonSvar.MISTET_JOBBEN -> JobbsituasjonBeskrivelse.HAR_BLITT_SAGT_OPP
                                DinSituasjonSvar.ALDRI_HATT_JOBB -> JobbsituasjonBeskrivelse.ALDRI_HATT_JOBB
                                DinSituasjonSvar.HAR_SAGT_OPP -> JobbsituasjonBeskrivelse.HAR_SAGT_OPP
                                DinSituasjonSvar.VIL_BYTTE_JOBB -> JobbsituasjonBeskrivelse.VIL_BYTTE_JOBB
                                DinSituasjonSvar.ER_PERMITTERT -> JobbsituasjonBeskrivelse.ER_PERMITTERT
                                DinSituasjonSvar.USIKKER_JOBBSITUASJON -> JobbsituasjonBeskrivelse.USIKKER_JOBBSITUASJON
                                DinSituasjonSvar.JOBB_OVER_2_AAR -> JobbsituasjonBeskrivelse.IKKE_VAERT_I_JOBB_SISTE_2_AAR
                                DinSituasjonSvar.VIL_FORTSETTE_I_JOBB -> null
                                DinSituasjonSvar.AKKURAT_FULLFORT_UTDANNING -> JobbsituasjonBeskrivelse.AKKURAT_FULLFORT_UTDANNING
                                DinSituasjonSvar.DELTIDSJOBB_VIL_MER -> JobbsituasjonBeskrivelse.DELTIDSJOBB_VIL_MER
                            }?.let { beskrivelse ->
                                JobbsituasjonMedDetaljer(
                                    beskrivelse = beskrivelse,
                                    detaljer = rs.getString(YRKESPRAKSIS)
                                        ?.takeIf { it.isNotBlank() && it != "X" && it != "-1" }
                                        ?.let {
                                            mapOf("stilling_styrk08" to it)
                                        } ?: emptyMap()
                                )
                            }
                        )
                    ),
                    annet = Annet(
                        andreForholdHindrerArbeid = when (AndreForholdSvar.valueOf(rs.getString(ANDRE_UTFORDRINGER))) {
                            AndreForholdSvar.JA -> JaNeiVetIkke.JA
                            AndreForholdSvar.NEI -> JaNeiVetIkke.NEI
                            AndreForholdSvar.INGEN_SVAR -> JaNeiVetIkke.VET_IKKE
                        }
                    ),
                    helse = Helse(
                        helsetilstandHindrerArbeid = when (HelseHinderSvar.valueOf(rs.getString(HAR_HELSEUTFORDRINGER))) {
                            HelseHinderSvar.JA -> JaNeiVetIkke.JA
                            HelseHinderSvar.NEI -> JaNeiVetIkke.NEI
                            HelseHinderSvar.INGEN_SVAR -> JaNeiVetIkke.VET_IKKE
                        }
                    )
                )
            )
        }

        private val noParams = emptyMap<String, Any>()

        private fun tilJson(obj: List<TekstForSporsmal>): String =
            try {
                ObjectMapper().writeValueAsString(obj)
            } catch (e: JsonProcessingException) {
                "[]"
            }

        private inline fun <reified T> readListOf(json: String?): List<T> =
            try {
                json?.let {
                    val type = mapper.typeFactory.constructParametricType(List::class.java, T::class.java)
                    jacksonObjectMapper().readValue<List<T>>(it, type)
                } ?: emptyList()
            } catch (e: IOException) {
                throw RuntimeException(e)
            }
    }
}